### DISCUSSION TITLE
Two behaviors of value template.

### DISCUSSION THREAD

**Discussion by ATaiIsHere** [sensedeep/dynamodb-onetable#410] [General]
I have a `MESSAGE` model to store messages for users.
I hope `GSI1PK`, `GSI1SK` are created when `isRead` is false, so I write my schema like this.
```
// schema
{
  version: '0.0.1',
  indexes: {
    'primary': { hash: 'pk', sort: 'sk' },
    'gsi-1': { hash: 'GSI1PK', sort: 'GSI1SK', project: 'all' },
  },
  models: {
    MESSAGE: {
      pk: { type: String, value: `USER#${userId}` },
      sk: { type: String, value: `MESSAGE#${id}` },
      GSI1PK: { type: String, value: `USER#${userId}` },
      GSI1SK: { type: String, value: true },
      GSI1SK: { type: String, value: true },
      userId: { type: String, required: true },
      id: { type: String, required: true, generate: 'ulid' },
      isRead: { type: Boolean, default: false },
    }
  },
};
```
```
// value
export default (
  model: AnyModel,
  fieldName: string,
  properties: OneProperties,
  params?: OneParams
): any => {
  if (model.name === 'MESSAGE') {
    if (fieldName === 'GSI1PK') {
      return properties?.isRead === false
        ? `USER#${properties.userId}`
        : properties?.isRead === true
        ? null
        : undefined;
    } else if (fieldName === 'GSI1SK') {
      return properties?.isRead === false
        ? `UNREAD`
        : properties?.isRead === true
        ? null
        : undefined;
    }
  }
  return undefined;
};
```
When use `gsi-1` to find message ( like `model.find({ GSI1PK: 'USER#1', GSI1SK: 'UNREAD' }, { index: 'gsi-1' })` )  ,
I got a `MissingError` as `Empty hash key. Check hash key and any value template variable references`,
> https://github.com/sensedeep/dynamodb-onetable/blob/main/src/Model.js#L1400
because it will call default function to build `GSI1PK`, `GSI1SK` anyway (line 1409).
but set `value` to a string template (line 1422), it will check `properties[name] === undefined` first.

why is there such a difference?

this makes me need to think more about I can use GSI1PK, GSI1SK to query directly or not.

**Comment by mobsense** [comment]
You need to provide the userId that is required by the value template for GSI1PK.

Also, if you put isRead in the value template, then the GSI entry would only get created when the value is true and you would not need to provide the GSI values to the model.find. Best practice is to use value templates to create the PK/SK, GSI PK/SK values for you.
